tests/test_cases/test_cocotb/test_queues.py

```diff
@@ -0,0 +1,191 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+"""
+Tests relating to cocotb.queue.Queue, cocotb.queue.LifoQueue, cocotb.queue.PriorityQueue
+"""
+from collections import deque
+
+import cocotb
+from cocotb.queue import Queue, PriorityQueue, LifoQueue, QueueFull, QueueEmpty
+from cocotb.regression import TestFactory
+
+
+async def run_queue_nonblocking_test(dut, queue_type):
+    queue = queue_type(10)
+
+    # queue empty
+    assert queue.maxsize == 10
+    assert queue.qsize() == 0
+    assert queue.empty()
+    assert not queue.full()
+
+    # put one item
+    queue.put_nowait(0)
+
+    assert queue.qsize() == 1
+    assert not queue.empty()
+    assert not queue.full()
+
+    # fill queue
+    if queue_type == PriorityQueue:
+        for k in range(9, 0, -1):
+            queue.put_nowait(k)
+    else:
+        for k in range(1, 10):
+            queue.put_nowait(k)
+
+    assert queue.qsize() == 10
+    assert not queue.empty()
+    assert queue.full()
+
+    # overflow
+    got_exception = False
+    try:
+        queue.put_nowait(100)
+    except QueueFull:
+        got_exception = True
+    assert got_exception
+
+    # check queue contents
+    if queue_type == LifoQueue:
+        for k in range(9, -1, -1):
+            assert queue.get_nowait() == k
+    else:
+        for k in range(10):
+            assert queue.get_nowait() == k
+
+    assert queue.qsize() == 0
+    assert queue.empty()
+    assert not queue.full()
+
+    # underflow
+    got_exception = False
+    try:
+        queue.get_nowait()
+    except QueueEmpty:
+        got_exception = True
+    assert got_exception
+
+    # task counts
+    assert queue._unfinished_tasks == 10
+
+    for k in range(10):
+        queue.task_done()
+
+    assert queue._unfinished_tasks == 0
+
+    got_exception = False
+    try:
+        queue.task_done()
+    except ValueError:
+        got_exception = True
+    assert got_exception
+
+
+factory = TestFactory(run_queue_nonblocking_test)
+factory.add_option("queue_type", [Queue, PriorityQueue, LifoQueue])
+factory.generate_tests()
+
+
+@cocotb.test()
+async def test_queue_contention_test(dut):
+    queue = Queue(10)
+
+    async def putter(queue, item):
+        await queue.put(item)
+
+    async def getter(queue, item):
+        assert item == await queue.get()
+        queue.task_done()
+
+    cr_list = deque()
+
+    # test put contention
+    for k in range(20):
+        cr_list.append(cocotb.fork(putter(queue, k)))
+
+    assert queue.qsize() == 10
+
+    # test killed putter
+    cr = cocotb.fork(putter(queue, 100))
+    cr.kill()
+    cr_list.append(cocotb.fork(putter(queue, 101)))
+
+    for k in range(20):
+        cr_list.append(cocotb.fork(getter(queue, k)))
+
+    cr_list.append(cocotb.fork(getter(queue, 101)))
+
+    while cr_list:
+        await cr_list.popleft().join()
```

Comment:
```suggestion
    await Combine(*cr_list)
```
You don't really need the `join()` either.